onnxruntime/contrib_ops/cpu/ngram.cc

```diff
@@ -0,0 +1,512 @@
+    case kTF: {
+      for (auto f : frequences) {
+        *output_data++ = static_cast<float>(f);
+      }
+    } break;
+    case kIDF: {
+      if (!w.empty()) {
+        assert(frequences.size() == w.size());
+        for (size_t i = 0; i < frequences.size(); ++i) {
+          *output_data++ = (frequences[i] > 0) ? w[i] : 0;
+        }
+      } else {
+        for (auto f : frequences) {
+          *output_data++ = (f > 0) ? 1.0f : 0;
+        }
+      }
+      break;
+      case kTFIDF: {
+        if (!w.empty()) {
+          assert(frequences.size() == w.size());
+          for (size_t i = 0; i < frequences.size(); ++i) {
+            *output_data++ = frequences[i] * w[i];
+          }
+        } else {
+          for (auto f : frequences) {
+            *output_data++ = static_cast<float>(f);
+          }
+        }
+      } break;
+      case kNone:  // fall-through
+      default:
+        assert(false);
+    }
+  }
+}
+
+template <typename T>
+void Ngram::ComputeImpl(OpKernelContext* ctx, size_t total_items) const {
+  const auto& impl = *impl_;
+  auto const set_end = impl.PoolEnd<T>();
+  // Frequency holder, init all to zero
+  std::vector<uint32_t> frequencies;
+  frequencies.resize(impl.output_size_, 0);
+
+  const auto N = impl.N_;
+  const auto S = impl.S_ + 1;  // Convert to distance
+  auto start_ngram_size = (impl.all_) ? impl.M_ : N;
+
+  auto X = ctx->Input<Tensor>(0);
+  auto const input_data = X->template Data<T>();
+  auto const end_data = input_data + total_items;
+  NGramItem<T> sample;
+
+  // Treat unigrams in a special way
+  if (start_ngram_size == 1) {
+    auto ngram_start = input_data;
+    while (ngram_start < end_data) {
+      sample.Clear();
+      sample.AddItem(*ngram_start);
+      ++ngram_start;
+      auto hit = impl.Find<T>(sample);
+      if (hit != set_end) {
+        // record frequency
+        auto ngram_id = hit->id();
+        impl.IncrementCount(ngram_id, frequencies);
+      }
+    }
+    if (++start_ngram_size > N) {
+      OutputResult(ctx, frequencies);
+      return;
+    }
+  }
+
+  // Convert skip into distance between n-gram items
+  // by adding 1
+  for (auto si = 1; si <= S; ++si) {
+    auto ngram_start = input_data;
+    while (ngram_start < end_data) {
+      // Check if any ni in [start_ngram_size..N]
+      // fit before end_data so we do not waste time adding [1..start_ngram_size)
+      // For that at least items with start_ngram_size should fit
+      auto at_least_this = ngram_start + (si - 1) * (start_ngram_size - 1) + (start_ngram_size - 1);
+      if (at_least_this >= end_data) {
+        break;
+      }
+      sample.Clear();
+      auto ngram_item = ngram_start;
+      for (auto ni = 1;
+           ni <= N &&
+           ngram_item < end_data;
+           ++ni, ngram_item += si) {
+        sample.AddItem(*ngram_item);
+
+        // Do not test anything before start_ngram_size
+        if (ni >= start_ngram_size) {
+          auto hit = impl.Find<T>(sample);
```

Comment:
I wonder what's best here, using one unorder_set  a hash to find the ngrams or a trie. The trie would optimize the finding of abc then bcd then cde… That's a probably a discussion for another PR.

Reply:
Yes, as discussed offline

Reply:
Yes, as discussed offline

---
In reply to: [242721399](https://github.com/Microsoft/onnxruntime/pull/180#discussion_r242721399) [](ancestors = 242721399)